Non-standard: Remove features included in ES6

webbedspace opened the pull request:
This removes the following items from the non-standard table which are
now in ECMAScript 6:
- function statement
- function "name" property (this has been added to the ES6 table)
- `__proto__`
- const, let
- RegExp "y" flag
- most of the String HTML methods
- String#substr
- octal

Closes #150.

webbedspace committed:
Non-standard: Removed features included in ES6

This removes the following items from the non-standard table which are
now in ECMAScript 6:
* function statement
* function "name" property (this has been added to the ES6 table)
* __proto__
* const, let
* RegExp "y" flag
* most of the String HTML methods
* octal

kangax commented:
@webbedspace Looks good, thanks! Didn't know about "substr" being in ES6 now... As far as "function statement", you mean "block-level function declaration" in ES6 table? I see that it's testing `{function f(){}} typeof f == "undefined"` and there's definitely at least couple more ways to test for it. `if (false) { function f() { } }` being one of them. In pre-ES6 envs. it would simply hoist to the top, IIRC. Also `if (true) { function f(){} } else { function f(){} }` and, finally, with "use strict".

kangax commented:
Btw, related tests by T.J. Crowder — http://jsbin.com/sedetiyo/1/edit (loose) and http://jsbin.com/sedetiyo/2/edit (strict)

webbedspace commented:
- Good point about block function hoisting. I guess I'll add an ES6 annex B test for the non-strict hoisting behaviour (given that it's in [B.3.3](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-block-level-function-declarations-web-legacy-compatibility-semantics).)
- [String#substr is actually in ES3.](http://www.ecma-international.org/publications/files/ECMA-ST-ARCH/ECMA-262,%203rd%20edition,%20December%201999.pdf) (back when Annex B was purely non-normative)

webbedspace commented:
Regarding those linked tests: test 1 is not adherent to the ES6 B.3.3 spec (which I think states that all block-level function declarations (BLFDs) are hoisted out in order of appearance, and, by implication, later bindings trump earlier bindings), so I wonder if you want a non-standard test for this particular block-level behaviour to be added. (Alas, I don't really have the resources to test that in most of the increasingly aged environs in the non-standard table - IE7, BESEN, etc.)

webbedspace committed:
ES6: Added hoisted block-level functions

This adds a test for the annex B behaviour for non-strict function
statements in blocks. I hope I've interpreted the spec correctly in this
case...

webbedspace committed:
Rebuilt the ES6 HTML file

webbedspace commented:
Also, octal was in ES3's annex B as well, so it's not been moved to the ES5 or ES6 table either.

(Even though annex B features prior to ES6 were not normative, and their implementation by platforms prior to ES6 is _arguably_ "non-standard", I've elected to only include annex B features on the ES6 list that were not mentioned in either ES3's annex B, or ES5's annex B, and thus are genuinely new to any version of the spec - even if, in the case of the string HTML methods, they predate even some of ES5's annex B features.)

kangax commented:
> test 1 is not adherent to the ES6 B.3.3 spec (which I think states that all block-level function declarations (BLFDs) are hoisted out in order of appearance, and, by implication, later bindings trump earlier bindings), [...]

I didn't even see that section on BLFD in ES6, since it was always disallowed in pre-ES6 times. Should probably update NFE article with that — http://kangax.github.io/nfe/#function-declarations-in-blocks

Thanks for bringing it up.

kangax commented:
*disallowed but still omnipresent as a de-facto, of course :)
